ksqldb-engine/src/main/java/io/confluent/ksql/engine/QueryIdUtil.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.engine;
+
+import com.google.common.collect.Iterables;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.metastore.model.DataSource.DataSourceType;
+import io.confluent.ksql.name.SourceName;
+import io.confluent.ksql.planner.plan.KsqlStructuredDataOutputNode;
+import io.confluent.ksql.planner.plan.OutputNode;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.query.id.QueryIdGenerator;
+import io.confluent.ksql.util.KsqlException;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Utility for constructing {@link QueryId}s - separate from {@code EngineExecutor} for
+ * easy access to unit testing.
+ */
+final class QueryIdUtil {
+
+  private QueryIdUtil() {
+  }
+
+  /**
+   * Builds a {@link QueryId} for a physical plan specification.
+   *
+   * @param metaStore   the meta store representing the current state of the engine
+   * @param idGenerator generates query ids
+   * @param outputNode  the logical plan
+   * @return the {@link QueryId} to be used
+   */
+  static QueryId buildId(
+      final MetaStore metaStore,
+      final QueryIdGenerator idGenerator,
+      final OutputNode outputNode
+  ) {
+    if (!outputNode.getSinkName().isPresent()) {
+      return new QueryId(String.valueOf(Math.abs(ThreadLocalRandom.current().nextLong())));
+    }
+
+    final KsqlStructuredDataOutputNode structured = (KsqlStructuredDataOutputNode) outputNode;
+    if (!structured.createInto()) {
+      return new QueryId("INSERTQUERY_" + idGenerator.getNext());
+    }
+
+    final SourceName sink = outputNode.getSinkName().get();
+    final Set<String> queriesForSink = metaStore.getQueriesWithSink(sink);
+    if (queriesForSink.size() > 1) {
+      throw new KsqlException("REPLACE for sink " + sink + " is not supported because there are "
```

Comment:
Why is there no check here on whether the sink specifies CREATE OR REPLACE or not?

Reply:
it'll actually fail before this point if it isn't a create or replace, but i'll make it explicit

Reply:
On second thought, I'm going to leave it like that - whether or not the statement is `CREATE OR REPLACE` is not the responsibility of the queryID generation. If it's attempting to create an already existing statement without `OR REPLACE` it should fail in the metastore (`putSource` specifically)